File: Algorithms/0540_Single_Element_in_a_Sorted_Array/Python/Single_Element_in_a_Sorted_Array_Solution_1.py
# Space: O(1)
# Time: O(logn)


class Solution:
    def singleNonDuplicate(self, nums):
        length = len(nums)
        if length == 0: return
        if length == 1: return nums[0]
        left, right = 0, length - 1
        while left < right:
            mid = (left + right) // 2
            mid_pair = mid + 1 if mid % 2 == 0 else mid - 1
            if nums[mid] == nums[mid_pair]:
                left = mid + 1
            else:
                right = mid
        return nums[left]


             

File: Algorithms/0540_Single_Element_in_a_Sorted_Array/Java/SingleElementinaSortedArraySolution1.java
// Space: O(1)
// Time: O(logn)

class Solution {
    public int singleNonDuplicate(int[] nums) {
        if (nums.length == 1) return nums[0];
        int left = 0, right = nums.length - 1;
        int mid, midPair;
        while (left < right) {
            mid = left + (right - left) / 2;
            midPair = mid % 2 == 0 ? mid + 1 : mid - 1;
            if (nums[mid] == nums[midPair]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return nums[left];
    }
}




